### PR TITLE
fix(server): audit correctness/perf/docs hardening + CLI arg validator (#288)

### PR DESCRIPTION
**Pull request by tbraun96**
## Summary

A multi-agent audit of `origin/main` (`f26f392f`) scanned the server for overflow/panic/resource-exhaustion bugs, docs/rustbook drift, and engine improvements; each finding was adversarially re-verified against the code before landing. This PR ships **all** of them, plus the **CLI arg validator** from #288.

Everything is verified: full workspace builds, **all 591 spark-server tests pass**, clippy clean, and the five previously-deferred items were verified **live** on a GB10 (dgx3) serving `nvidia/Qwen3.6-35B-A3B-NVFP4`.

## Correctness / security fixes

| Area | Bug | Fix |
|---|---|---|
| Prefill budget | `max_tokens - 1` underflows to `usize::MAX` (release) / panics (debug) on a `max_tokens:0` scoring call that samples a spontaneous `<think>` token | `saturating_sub(1)` at all 3 promotion sites |
| Chat `n` | Blocking `/v1/chat/completions` read `n` unbounded → alloc/compute DoS | Reject `n∉[1,128]`; fixes a `REQUESTS_ACTIVE` gauge leak on validation 400s too |
| Completions params | `temperature`/penalties passed to the sampler with no range check | Range-validate → 400, mirroring chat |
| Tool parser | Two `&s[..n]` byte slices on raw model output panic on a multibyte char | Char-boundary truncation |
| EP broadcast | `ep_broadcast_tokens` staged the full prompt into chunk-sized scratch → device OOB write | Bound-check against `scratch_bytes` |
| MTP gate | Re-measure read SSM state before the secondary-stream restore landed | `sync_secondary()` + clear drafts at the `MeasureDecode` arm |
| **Swap-out slot double-own** | `swap_out_sequence` compacted a survivor onto the position index (double-own → cross-request GDN state bleed) and overwrote the victim's slot before the save; ignored EP v2 | **Extract retire's proven two-phase exclusive-claim compaction into a shared `compact_survivors_into_range`; reorder swap-out to save → free victim → recompact. Skipped under EP v2.** |

## Performance

- **Streaming stop-string scan** O(n²)→O(n) (bounded straddling-suffix search).
- **Incremental detokenizer** O(n²)→O(n): the chat + completions streaming paths re-decoded the entire token history every token. New `ChatTokenizer::incremental_decode` (vLLM `detokenize_incrementally`) decodes only the suffix window. **Verified byte-identical** to full-decode across an 8-case corpus (ASCII, TOML/code, CJK, emoji + ZWJ family sequences, mixed Cyrillic/Japanese/Korean/Arabic) using the server's own `/tokenize`+`/detokenize`.
- **Per-token buffer reuse** on the host sampling path: the `n·vocab·elem` staging buffer and the per-seq `f32` logits buffer are now per-thread thread-local scratch (fully overwritten each use → byte-identical). Verified live via temp>0 + grammar tool-call output.

## Docs / rustbook

- Corrected stale CLI defaults (`--max-seq-len` 4096→32768, `--mtp-quantization` nvfp4→bf16, `--tool-max-tokens` 512→8192, `--default-min-p` 0.0→0.08), auth model, rate-limiter defaults, tool-parser tables, README/QUICKSTART K=2-vs-K=3 mislabel + wrong parser, CHANGELOG (#170→#180 + 5 missing entries), GpuBackend method count, DeepSeek-V4 in the model matrix, and single-file→module-directory paths.
- **Systemic stale-link sweep** (~35 refs across the book + `AGENTS.md`): `docs/design/*`→`docs/adr/*` + top-level docs, `docs/history/*`→`docs/ATLAS_SPARK_JOURNEY.md`, the nonexistent `moe_dispatch.cu`→`layers/moe/forward_ep.rs` + `moe_w4a16_grouped_gemm.cu`, `scripts/bench-concurrency.sh`→`scripts/run_conc_benchmark.sh`, `CONCURRENCY_SWEEP_RESULTS.md`→`bench/`. Every replacement path verified to exist.

## CLI arg validator (#288)

New `cli::validate_serve_args` — a single place that turns contradictory flag combinations into **hard errors** (never warnings), collected and reported together, each with **what / why / fix** shaped for humans and AI agents, run **before** the multi-minute model load. Rules cover the issue's example (`--fp8-kv-calibration-tokens` without `--kv-cache-dtype fp8`) plus: `--kv-high-precision-layers` on a bf16 cache, `--require-auth` with no token, `--num-drafts` without a speculative method, `--disable-thinking` + thinking budget, `rank`/`ep-size`/`tp-size` vs `world-size`, orphaned high-speed-swap options, `gpu-memory-utilization` range, and enum-value typos (with the valid set listed). Unit-tested (11 tests).

Example output:
```
Atlas CLI: 2 invalid flag combinations — fix before serving:

  [1] --fp8-kv-calibration-tokens 256 has no effect with --kv-cache-dtype bf16.
      why: online FP8 KV-scale calibration only feeds an FP8 KV cache; with a
           bf16/nvfp4 cache the calibrated scales are never read.
      fix: set --kv-cache-dtype fp8, or drop --fp8-kv-calibration-tokens (0 = off).

  [2] --require-auth is set but no bearer tokens were provided.
      why: with auth enforced and no tokens loaded, EVERY request is rejected 401.
      fix: pass --auth-tokens-file <path> (preferred, 0600) or --auth-token <token>.
```

## Live verification (dgx3, Qwen3.6-35B-A3B-NVFP4)

- **Detok byte-identity**: 8/8 corpus cases byte-identical (incremental == full decode), incl. CJK/emoji/ZWJ.
- **Slowloris**: a header-trickle connection is closed at exactly 30.0s; normal requests work; a 45.8s generation completes uninterrupted (a blanket request-timeout would have wrongly killed it).
- **Host-path (F36/F37)**: temp=0.8 haiku + grammar tool-call produce correct output.
- **Swap-out refactor**: 16 concurrent unique-secret requests under `ATLAS_PREFILL_CODISPATCH=1` → 16/16 own-code-correct, **0 cross-request bleed**, 0 errors, 0 crashes (exercises the shared `compact_survivors_into_range` via retire). Note: forcing an actual disk swap-out needs impractical concurrency here (the KV pool is weight-dominated large), so the swap-out trigger itself was not hit live; the fix is a faithful reuse of the already-proven retire-path logic and the shared function is exercised under co-dispatch.

🔬 Found via multi-agent audit; every finding independently re-verified, and the deferred items verified live before landing.

### CODE REVIEW

**Comment by SeedSource** [comment]
## [BETA] Automated PR review — SeedSource AI pipeline

_Advisory only, not a merge gate. AI-generated (internal risk-triage, Atlas serve gate, GLM-5.2 review), attributed per Atlas convention._

### Triage
- Risk: **low** - serve-test-required (triage advisory): `False` - path-heuristic (authority): `True`
- Rationale: Documentation and changelog updates only; no code changes affecting inference or quantization logic.
- Serving-path files: `crates/spark-model/src/model/impl_a2.rs, crates/spark-server/src/api/chat/mod.rs, crates/spark-server/src/api/chat_phases.rs, crates/spark-server/src/api/chat_stream/handle_token.rs, crates/spark-server/src/api/chat_stream/state.rs, crates/spark-server/src/api/completions.rs`

### Serve gate
- Build: 34s - Qwen3.6-35B-A3B-FP8 on GB10 node1
- Top-k KL vs main: **0.0010** (no shift vs main) - decode delta **-0.3%**
- Coherence probe: coherence=pass ping=10/10 _(note: our test_coherence.py is flaky here; KL is the reliable serve signal)_

### GLM-5.2 code review
**Verdict:** Approve with minor follow-up.

**Scope:** This is a docs-only PR — updating references from `docs/design/` and `docs/history/` to the new `docs/adr/` structure and `ATLAS_SPARK_JOURNEY.md`. No code changes.

**Correctness:** The doc-path migrations are internally consistent across AGENTS.md, book chapters, and reading references. The `api.rs` → `api/` and `scheduler.rs` → `scheduler/` updates suggest a module split that should be verified to actually exist in the codebase (diff is truncated, so I can't confirm). The CHANGELOG entries (#293, #291, #257, #278) reference prior merged work — no risk here.

**Concerns:**
1. **Broken link risk:** Multiple new `docs/adr/NNNN-*.md` paths are cited (0004, 0003, 0007, 0011, 0010). These files must exist post-merge or the book build breaks. CI should catch this if mdbook link-checking is configured.
2. **QUICKSTART.md tool-parser change:** `hermes` → `qwen3_coder` in the docker recipe is a behavioral change for users copy-pasting. The prose above it still lists both options, so this is intentional, but it's worth a second look — is `qwen3_coder` correct for the default model in that recipe?
3. **README `--num-drafts` removal:** The explanation paragraph is clear and well-reasoned. Low risk.

**Risk:** Low — docs only. No perf/correctness impact.

<sub>Pipeline: build+serve on GB10, GLM-5.2 via NVIDIA NIM, internal risk-triage classifier. External diff treated as untrusted input. Public repo.</sub>

**Comment by tbraun96** [comment]
## Benchmark gate results — Gate A: ✅ PASS

Ran `/benchmark-pr` on **dgx3** against `atlas-gb10:pr298` (branch SHA `67d71659`), 35B FP8 flagship, exact PR #278 webserver_ok recipe (`Qwen/Qwen3.6-35B-A3B-FP8`, `--kv-cache-dtype bf16 --lm-head-dtype nvfp4 --speculative --num-drafts 1 --mtp-quantization bf16 --enable-prefix-caching --ssm-cache-slots 256 --kv-high-precision-layers auto`).

| Metric | Result | Gate |
|---|---|---|
| `webserver_ok` | **10 / 10** | 10/10 ✅ |
| `followed_directions` | **10 / 10** | 10/10 ✅ |
| **Σ(wall_time_s)** | **919.9 s** | ≤ 1300 ✅ |

Per-run walls (s): `73.1  105.0  73.0  86.9  78.6  73.5  184.7  80.8  70.2  94.1` — median **80.8**, min **70.2**, one 184.7 outlier still leaves us **380 s under budget**. That's the lowest Σwall I've recorded on this recipe (vs the prior ~1364 s that first cleared the <1500 target on the SSM tail-protect work) — the incremental-detokenizer (O(n²)→O(n)) and the per-token buffer reuse landing on the decode/emission hot path show up here.

### The gate earned its keep 🛡️
It caught a **ship-blocking false positive** in the new CLI arg validator (#288) before this ran: the validator rejected `--kv-cache-dtype bf16 --kv-high-precision-layers auto` — which is exactly the canonical flagship serve recipe. Fixed in `67d71659` (redundant ≠ invalid), with a regression test asserting the flagship recipe validates. Without the benchmark, that would have broken every operator running the standard command.

**Gate B (subset ST-995 BFCL, vs baseline 84.66 / 83.32) is running now** — will post when it lands. Not merging; that's your call.

**Comment by tbraun96** [comment]
## Gate B + C results — a real bug caught, root-caused, and fixed ✅

Running `/benchmark-pr` on dgx3 caught a genuine bug in this PR and I've now fixed it.

### The bug: greedy decoding became non-deterministic
BFCL (35B FP8, temp=0, seed=42, non-streaming) on the PR was **non-deterministic**:

| Run | overall / normalized | per-sample |
|---|---|---|
| **Base** `f26f392f` (×2 builds) | **84.66 / 83.32** | **0/1004 diffs — deterministic** |
| PR `67d71659` run 1 | 84.56 / 83.23 | — |
| PR `67d71659` run 2 | 84.26 / 82.94 | 21/1004 differ vs run 1 |

Two greedy runs of the PR disagreed on 21/1004 samples (borderline-argmax flips), while base reproduces to the byte. So the PR broke temp=0/seed=42 reproducibility — worse than a fixed dip.

### Root cause: the MTP-gate re-measure sync (item 4 of this PR)
The `sync_secondary()` + `pending_drafts.clear()` I added to the `GateStep::MeasureDecode` arm. In a varying-prompt-length workload the throughput gate **re-measures on nearly every request** (the depth regime flips each sample: 4233→1195→2430→5239…), so that hunk ran constantly and reordered the async SSM live-state restore against default-stream compute inside the timing-derived measurement branch — making the shared SSM recurrent state depend on run-to-run GPU timing.

Root-caused via a multi-agent workflow (3/3 adversarial verifiers) + an empirical control leg. F36/F37 buffer reuse were **cleared** — `copy_d2h` is synchronous (`cuMemcpyDtoHAsync_v2` + `cuStreamSynchronize`), so they're provably byte-identical to base.

### The fix (`7d10585c`)
Revert just that hunk → `scheduler/mod.rs` is now **byte-identical to base**; F36/F37 kept. A control image built from exactly this fix scored:

| | overall / normalized | vs base |
|---|---|---|
| **Fixed PR** (`7d10585c`) | **84.66 / 83.32** | **0/1004 diffs — identical to base** |

Byte-identical to the deterministic base ⇒ determinism restored, Gate B passes (≥ baseline).

The stale-read that hunk targeted is a non-BFCL path; if genuinely needed it belongs on the MTP→decode boundary unconditionally (as the batch path already does), not inside the timing-dependent gate-measurement branch — tracked as a separate, benchmark-gated change.

### Final gate scorecard (dgx3, `atlas-gb10:pr298` → fix `7d10585c`)
| Gate | Result |
|---|---|
| **A** — webserver_ok N=10 | ✅ **PASS** — 10/10, 10/10, Σwall **919.9s** |
| **B** — BFCL subset ST-995 | ✅ **PASS** (after fix) — **84.66 / 83.32** = base, deterministic |
| **C** — warm-TTFT | ✅ **PASS** — 264.7/266.7 ms vs base 265.0/267.2 (−0.11%/−0.19%) |

All three gates pass with the fix. Not merging — your call.
